libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2164,6 +2164,52 @@ void TypeChecker::typeCheckFunctionGeneralChecks(
 			m_errorReporter.typeError(errorId, paramArgMap[i]->location(), description);
 		}
 	}
+
+	TypePointers const& returnParameterTypes = _functionType->returnParameterTypes();
+	bool isLibraryCall = (_functionType->kind() == FunctionType::Kind::DelegateCall);
+	bool callRequiresABIEncoding =
+		// ABIEncode/ABIDecode calls not included because they should have been already validated
+		// at this point and they have variadic arguments so they need special handling.
+		_functionType->kind() == FunctionType::Kind::DelegateCall ||
+		_functionType->kind() == FunctionType::Kind::External ||
+		_functionType->kind() == FunctionType::Kind::Creation ||
+		_functionType->kind() == FunctionType::Kind::Event;
+
+	if (callRequiresABIEncoding && !experimentalFeatureActive(ExperimentalFeature::ABIEncoderV2))
+	{
+		solAssert(!isVariadic, "");
+		solAssert(parameterTypes.size() == arguments.size(), "");
+		solAssert(!_functionType->isBareCall(), "");
+		solAssert(*_functionCall.annotation().kind == FunctionCallKind::FunctionCall, "");
+
+		for (size_t i = 0; i < arguments.size(); ++i)
+		{
+			solAssert(parameterTypes[i], "");
+
+			if (!typeSupportedByOldABIEncoder(*parameterTypes[i], isLibraryCall))
+				m_errorReporter.typeError(
+					2443_error,
+					paramArgMap[i]->location(),
+					"The type of this parameter, " + parameterTypes[i]->toString(true) + ", "
+					"is only supported in ABIEncoderV2. "
+					"Use \"pragma experimental ABIEncoderV2;\" to enable the feature."
+				);
+		}
+
+		for (size_t i = 0; i < returnParameterTypes.size(); ++i)
+		{
+			solAssert(returnParameterTypes[i], "");
+
+			if (!typeSupportedByOldABIEncoder(*returnParameterTypes[i], isLibraryCall))
```

Comment:
You mean that I should add a check for `EVMVersion.supportsReturndata()` and if it's unsupported and the result is a dynamic type, report an error even with encoder v2?

Reply:
I don't have a solution, but that could be another complication. It can be argued, though, that we don't really have to support these old EVMs too much anymore...

Reply:
I checked dynamic arrays on older EVM and looks like we're covered. I have actually fixed a bug in that code recently (#9815) so I should have realized right away. The code generator has a check for return data support and just returns an inaccessible dynamic for stuff that's not supported by the EVM. So it will be handled correctly at runtime and if you try to use the result you'll get a compilation error too:

`nested-dynamic-v2.sol`:
``` solidity
pragma experimental ABIEncoderV2;

contract C {
    function get() public view returns (uint[][] memory) {}

    function test() public view returns (bool) {
        uint[][] memory x = this.get();
    }
}
```

``` bash
solc nested-dynamic-v2.sol --evm-version homestead
```
``` solidity
Error: Type inaccessible dynamic type is not implicitly convertible to expected type uint256[] memory[] memory.
 --> nested-dynamic-v2.sol:7:9:
  |
7 |         uint[][] memory x = this.get();
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Reply:
ok great!

Reply:
I added this as a test case (actually two - one for homestead and one for byzantium and later).